### ISSUE TITLE
Unable to create lock in backend with simultaneous backups from different hosts?

### ISSUE THREAD

**Issue by mrmachine** [restic/restic#2059]
Output of `restic version`
--------------------------

restic 0.9.3 compiled with go1.11.1 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

If lock is required but cannot be acquired because another host/pid has the lock, restic should have an option to wait and retry until either the lock can be acquired or some timeout expires or retry count is exceeded.

What are you trying to do?
--------------------------

Trying to use https://github.com/ixc/restic-pg-dump on multiple hosts with the same (default) cron schedule (`0 */4 * * *`).

Sometimes, depending on the specific timing (how long it takes to do a database dump, how long it takes one host or the other to send data to S3), it seems I get an error like this:

```
Fatal: unable to create lock in backend: repository is already locked exclusively by PID 283 on 24e1b7101987 by root (UID 0, GID 0)
```

Did restic help you or made you happy in any way?
-------------------------------------------------

Yes! It's amazing to find a feature rich and active open source project offering encrypted incremental and network efficient backups that is easy to use.

In particular, the ability to thin the repository to N latest, daily, weekly, etc. with deduplication is exactly what I needed. Thanks!

**Comment by fd0** [comment]
Hm, thanks for the report. In my opinion, that's a shortcoming of the scripts in the docker container that you're using. There are two types of locks in a repository: exclusive (used for prune and forget), and non-exclusive (all other operations). There may be more than one non-exclusive lock (so multiple instances of restic can save their data simultaneously) but at most one exclusive lock. When an exclusive lock exists, no other locks must exist in parallel, even non-exclusive locks are forbidden.

In practice, this means that you can run `restic backup` from several different hosts and use the same repo, but when `prune` is running, no backups can be made until it finishes.

In general, running `restic forget` after each `restic backup` operation is okay, but running `prune` (via `restic forget --prune`) will block the repo. That's what you're experiencing.

I suggest you think about what you schedule is: If you have a daily backup for several machines, run `prune` only once per week (or so) while no other machines make backups.

This is not an issue in restic, so I'm closing this issue for now. Please feel free to add further comments. Thanks!

**Comment by mrmachine** [comment]
@fd0 thanks for clarifying the way restic works. However, I still think this feature request is valid. If I am attempting a backup, but some other scheduled process/server (or interactive manually triggered process) is doing `forget --prune` at the same time, it will fail with an error.

Even if that only happens once a week, instead of 50% of the time every 4 hours, it's still a missed backup. And I expect that the longer I delay in doing a prune, or the bigger the backup size, the longer the prune will take, and the more servers are backing up to or the more people who are using the same repository, will increase the chance that backups will fail during that window.

I still believe that restic should (and could without too much difficulty) expose an option to retry (up to a retry limit or timeout) when it is unable to acquire a lock, before failing with an error. For example, `--retry 10` (10 times) or `-- retry 5m` (keep retrying for 5 minutes).

The only alternative I can see is for everyone attempting to schedule backups and wanting reliable backups is to re-implement their own retry logic on failure to acquire a lock, which seems like something everyone scheduling backups would need to do (repetitive).

Even if I were to do as you suggest, it's not necessarily within my control to ensure that every developer who deploys this to any/every server will not configure an overlapping schedule. And it's a pain to have to run not only a scheduled task on every server, but also run one "master" scheduled task to run prune on a non-overlapping schedule.

I think that significantly increases the setup complexity required to get backups up and running using an image like this. It's no longer just a one liner `docker run ...` to get scheduled automated backups.

**Comment by mholt** [comment]
Hmmm, I disagree; I think that is feature creep, and complexity that restic doesn't need to incur. The bug is in the script that is running it at the wrong time, and retries are trivial in scripts/wrapping programs anyway.

**Comment by dimejo** [comment]
> Hmmm, I disagree; I think that is feature creep, and complexity that restic doesn't need to incur. The bug is in the script that is running it at the wrong time, and retries are trivial in scripts/wrapping programs anyway.

I disagree. :)

IMHO feature creep is to add features which are rarely needed. Having some retry mechanic probably is useful for most people. Do you expect all of them to script something on their own? What about people who aren't able to do that? Keep in mind that not everyone is a programmer.

I've wrote a small wrapper script for restic to handle everything on my Unix servers. But while I do backups on Windows aswell my batch script skills are close to non existant, hence there is no retry logic for my backups on Windows. If a backup is missed because prune or check haven't finished it'll be executed on the next run 24 hours later.

I'm aware that there are issues/features which are more important than this one, but think this a valid RFE and the issue should atleast be reopened to keep track of it.

**Comment by mrmachine** [comment]
Any chance we can reopen this? Backups can also fail unexpectedly when a user is running an unscheduled `forget --prune`. Surely not all usage of restic is carefully scheduled and scripted by programmers?

**Comment by darkl0rd** [comment]
A "wait" option would definitely be beneficial, at the very least restic could return a > 1 return code when the repository is exclusively locked for sake of (easier) scripting purposes.

My scenario is similar to what is described in posts above; I run a backup + forget every hour and perform a prune once a week. This prune takes well over an hour and therefore results in scheduled backups failing.

Right now, I catch this by checking the return code of restic, and 'grepping' for "exclusively locked" - when that is the case, I back-off for 5 minutes and try again. This would be easier if restic would return something else than 1 (which it already returns for all other errors), but it would most definately be even more convenient if restic where able to just wait itself.

Alternatively, would it be possible to not require an exclusive lock when performing a prune?
I'm not privy to the internals, but I'm imagining that 'forget' marks the datasets as no longer in use; shouldn't prune be able to remove the marked datasets without requiring an exclusive lock in the first place?

**Comment by mrmachine** [comment]
Just as an additional data point, it's just taken me a few hours to update https://github.com/ixc/restic-pg-dump to retry when it encounters exclusive locks (actually when it encounters any error, which might not be appropriate). And my first attempts were buggy and I didn't realise I ended up with many overlapping/waiting restic scripts (triggered by cron). And I had to put this retry handling code around basically every call to `restic` (backup, check, forget, prune).

In searching restic issues on GitHub it appears that there have been quite a few similar reports in the past about issues when backing up from multiple hosts. I can't imagine that scenario is that uncommon, so I still really think it's a valid request for restic to implement it's own wait/retry option.

If the goal is reliable backups, and multiple hosts is a supported use case, then I think this is really essential to save users from implementing this time consuming and error prone task themselves, over and over again.
